style: sync changes from mozilla-central.

emilio opened the pull request.

dholbert committed:
style: Remove pref for CSS Containment (layout.css.contain.enabled) r=AlaskanEmily

(Since we've been shipping with it default-enabled for a while now.)

See https://bugzilla.mozilla.org/show_bug.cgi?id=1466008#c9 through #c13 for
notes on the reftest.list change.

Differential Revision: https://phabricator.services.mozilla.com/D71861

emilio committed:
style: Keep track of nested dependencies for :where() and :is().

The tricky part of :is() and :where() is that they can have combinators inside,
so something like this is valid:

  foo:is(#bar > .baz) ~ taz

The current invalidation logic is based on the assumption that you can
represent a combinator as a (selector, offset) tuple, which are stored in the
Dependency struct. This assumption breaks with :is() and :where(), so we need
to make them be able to represent a combinator in an "inner" selector.

For this purpose, we add a `parent` dependency. With it, when invalidating
inside the `:is()` we can represent combinators inside as a stack.

The basic idea is that, for the example above, when an id of "bar" is added or
removed, we'd find a dependency like:

    Dependency {
        selector: #bar > .baz,
        offset: 1, // pointing to the `>` combinator
        parent: Some(Dependency {
            selector: foo:is(#bar > .baz) > taz,
            offset: 1, // Pointing to the `~` combinator.
            parent: None,
        })
    }

That way, we'd start matching at the element that changed, towards the right,
and if we find an element that matches .baz, instead of invalidating that
element, we'd look at the parent dependency, then double-check that the whole
left-hand-side of the selector (foo:is(#bar > .baz)) actually changed, and then
keep invalidating to the right using the parent dependency as usual.

This patch only builds the data structure and keeps the code compiling, the
actual invalidation work will come in a following patch.

Differential Revision: https://phabricator.services.mozilla.com/D71421

emilio committed:
style: Make Invalidation work in terms of a dependency, not a selector.

That way we can look at the parent dependency as described in the previous
patch. An alternative would be to add a:

    parent_dependency: Option<&'a Dependency>

on construction to `Invalidation`, but this way seems slightly better to avoid
growing the struct. It's not even one more indirection because the selector is
contained directly in the Dependency struct.

Differential Revision: https://phabricator.services.mozilla.com/D71422

emilio committed:
style: Optimize invalidation by scanning the rightmost compound inside :where() and :is() with the outer visitor.

See the comment about why this is valuable. For a selector like:

    .foo:is(.bar) > .baz

Before this patch we'd generate an Dependency for .bar like this:

    Dependency {
        selector: .bar,
        offset: 0,
        parent: Some(Dependency {
            selector: .foo:is(.bar) > .baz,
            offset: 1, // Pointing to the `>` combinator.
            parent: None,
        }),
    }

After this patch we'd generate just:

    Dependency {
        selector: .foo:is(.bar) > .baz,
        offset: 1, // Pointing to the `>` combinator.
        parent: None,
    }

This is not only less memory but also less work. The reason for that is that,
before this patch, when .bar changes, we'd look the dependency, and see there's
a parent, and then scan that, so we'd match `.bar` two times, one for the
initial dependency, and one for .foo:is(.bar).

Instead, with this we'd only check `.foo:is(.bar)` once.

Differential Revision: https://phabricator.services.mozilla.com/D71423

emilio committed:
style: Handle disjoint selectors in the selector map.

This way, something like:

  *:where(.foo, .bar)

Will end up twice on the selector map, just as if you would've written
.foo, .bar.

But we're a bit careful to not be wasteful, so:

  .foo:where(div, span)

Will still end up using the .foo bucket.

It needs a bit of borrow-checker gymnastics to avoid cloning the entry
in the common path. It's a bit gross but not too terrible I think.

Differential Revision: https://phabricator.services.mozilla.com/D71457

emilio committed:
style: Collect ancestor hashes from single-length :is and :where selector lists.

We can only collect hashes from single-length selectors, as described in
the comment.

Differential Revision: https://phabricator.services.mozilla.com/D71458

Commit:
style: Implement style system support for Masonry layout.

This implements support for this CSS Masonry layout proposal:
https://github.com/w3c/csswg-drafts/issues/4650

I've intentionally left out a shorthand (place-tracks?) for now until
we have a draft CSS spec for this.

Differential Revision: https://phabricator.services.mozilla.com/D67061

emilio committed:
style: Only override to default color in high-contrast / forced-colors mode if inheriting from transparent.

That way elements inside links, form controls, etc have the right
contrast, even if the page overrides the color.

We can't do it when inheriting from transparent because we've already
forgotten about the "right" color to inherit, so the default color makes
sense. But that is a pretty unlikely edge case.

Differential Revision: https://phabricator.services.mozilla.com/D73069

BorisChiou committed:
style: Add ::marker when checking may_have_animations() during traversal.

When unhidding a ::marker element, we construct its generated item, and
then call StyleNewSubtree() on this generated item. During traversal, we
may update any animation related values in Gecko_UpdateAnimations(), which
may update the base styles for animation properties.

The test case is an animation segment from "null" to "inital" value. We
replace the "null" value with the base style value for the specific animation
property, so we can do interpolation properly.
(e.g. opacity: "null => initial" becomes "1.0 => initial")
If we don't update the animation related values in
Gecko_UpdateAnimations after generating ::marker, we may do
interpolation from "null" to "initial", which causes a panic.

Differential Revision: https://phabricator.services.mozilla.com/D73408

sigiesec committed:
style: Make NotNull move its member pointer where possible.

Differential Revision: https://phabricator.services.mozilla.com/D72827

sigiesec committed:
style: Improve handling of copying arrays in layout/style/ and servo/.

Differential Revision: https://phabricator.services.mozilla.com/D72351

AlaskanEmily committed:
style: Use fill length rather than index to indicate a repeat(auto) in subgrid from Servo.

Differential Revision: https://phabricator.services.mozilla.com/D70066

lokalmatador committed:
style: Convert style-font #defines to an enum class.

Differential Revision: https://phabricator.services.mozilla.com/D73728

heycam committed:
style: Gracefully handle errors creating shared memory UA style sheets.

We still panic in a debug build, so that developers can notice when they
need to add a new static atom after modifying UA sheets.

We also add telemetry to note when this happens, add an app note to a
crash report, in case any crash later on occurs, and re-up the existing,
expired shared memory sheet telemetry probes so we can look at them
again.

Differential Revision: https://phabricator.services.mozilla.com/D73188

emilio committed:
style: Clean up cascade rule iteration. r=nordzilla

The current API was pretty awkward as a result of two things:

 * Not being able to create empty iterators for smallbitvec.
 * We used to call the `F` function multiple times, but turns out that
   collecting the declarations in a SmallVec was a perf win.

So clean this up so that it looks more similar to other APIs, taking an
iterator directly.

This is a bit more code, but hopefully easier to understand (and also hopefully
easier to optimize).

The motivation for this work is that I plan to investigate rebasing / landing
https://github.com/servo/servo/pull/20151, and I don't want more instantiations
of apply_declarations and such.

Differential Revision: https://phabricator.services.mozilla.com/D74369

longsonr committed:
style: Don't apply minimum font sizes to SVG text.

Differential Revision: https://phabricator.services.mozilla.com/D74581

emilio committed:
style: Unprefix -moz-read-write / -moz-read-only.

And remove some duplicated tests from WPT.

Differential Revision: https://phabricator.services.mozilla.com/D75231

emilio committed:
style: Make ::-moz-focus-outer a no-op, and remove it on Nightly.

See https://bugzilla.mozilla.org/show_bug.cgi?id=932410#c2 for the
context for which this pseudo-element was added.

In the previous patch, I had to special-case range appearance because of
this pseudo-class, but that patch makes this pseudo-class completely
redundant, as now all form controls, themed and unthemed, display
outlines, unless the native theme displays a focus indicator on its own.

Remove the special case, and make ranges use outlines like everything
else rather than this bespoke pseudo-element.

Differential Revision: https://phabricator.services.mozilla.com/D74734

emilio committed:
style: Fix parsing of :is() and :where() to account for constraints from parent selectors.

Differential Revision: https://phabricator.services.mozilla.com/D75856

BorisChiou committed:
style: Replace AspectRatio with computed::position::Ratio in media-queries.

Also, we drop the pref, layout.css.aspect-ratio-number.enabled, becacuse
the spec of css-sizing-4 uses Number now.

Differential Revision: https://phabricator.services.mozilla.com/D75233

BorisChiou committed:
style: Use style::One for Integer to avoid implementing Mul.

Differential Revision: https://phabricator.services.mozilla.com/D76207

emilio committed:
style: Fix a no-longer valid assumption in pseudo-element matching / invalidation code.

After bug 1632647, we can have pseudo-classes inside :not / :is /
:where, which the invalidation and matching code weren't handling.

Add a few tests for this stuff working as expected.

Differential Revision: https://phabricator.services.mozilla.com/D76160

emilio committed:
style: Fix a case where we'd allow parsing functional :host incorrectly.

This is a missing check I should've introduced in bug 1632647.

Differential Revision: https://phabricator.services.mozilla.com/D76161

BorisChiou committed:
style: Let aspect-ratio (css-sizing-4) support 'auto | <ratio>'.

In order to test its parsing and serialization, we expose it but protect
it behind a pref.

Besides, I would like to drop layout.css.aspect-ratio-number.enabled in
the next patch because the spec has been updated. It seems we don't have
to keep this pref and we should always use Number.

Differential Revision: https://phabricator.services.mozilla.com/D74955

emilio committed:
style: Enable :is() and :where() in UA sheets.

This will allow us to clean them up.

Differential Revision: https://phabricator.services.mozilla.com/D76262

emilio committed:
style: Support field_bound in #[derive(Parse)].

Differential Revision: https://phabricator.services.mozilla.com/D76268

emilio committed:
style: Clean up parsing of UnicodeRange.

Differential Revision: https://phabricator.services.mozilla.com/D76330

emilio committed:
style: Derive parse for ShapeRadius.

Depends on D76331

Differential Revision: https://phabricator.services.mozilla.com/D76332

emilio committed:
style: Derive parse for TextOverflowSide.

Depends on D76330

Differential Revision: https://phabricator.services.mozilla.com/D76331

Commit:
style: Allow 'opacity' on ::first-letter/::first-line pseudos.

Differential Revision: https://phabricator.services.mozilla.com/D76387

sefeng211 committed:
style: Push/Pop dialog to top layer when needed.

This patch completes the top layer requirement for showModal()
Spec: https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-showmodal

Differential Revision: https://phabricator.services.mozilla.com/D74922

emilio committed:
style: Implement the ::file-chooser-button pseudo-element.

As per https://github.com/w3c/csswg-drafts/issues/5049.

Don't enable it unconditionally just yet, as the name may change.

I had to move some rules in forms.css because otherwise you get
specificity conflicts.

Differential Revision: https://phabricator.services.mozilla.com/D76214

emilio committed:
style: Remove the @-moz-document url-prefix() hack preference, enable it everywhere.

It doesn't seem like realistically we're going to be able to get rid of
this any time soon.

Differential Revision: https://phabricator.services.mozilla.com/D76809

emilio committed:
style: Remove two useless mem::replace calls.

Differential Revision: https://phabricator.services.mozilla.com/D76884

emilio committed:
style: Manually tweak inlining in stateless pseudo selector matching.

This addresses a minor regression in bloom-matching.html. The common
case here is that there's no selector to the right of the
pseudo-element, so keep that path inline, while keeping all other checks
out of line.

Differential Revision: https://phabricator.services.mozilla.com/D76793

emilio committed:
style: Finer grained invalidation for attribute changes.

This should help out quite a bit with uBO, which has lots of very
general attribute selectors. We invalidate per attribute name rather
than using a SelectorMap, which prevents matching for attribute
selectors that can't have changed.

The idea is that this should be generally cheaper, though there are
cases where this would be a slight pesimization. For example, if there's
an attribute selector like:

  my-specific-element[my-attribute] { /* ... */ }

And you change `my-attribute` in an element that isn't a
`my-specific-element`, before that the SelectorMap would've prevented us
from selector-matching completely. Now we'd still run selector-matching
for that (though the matching would be pretty cheap).

However I think this should speed up things generally, let's see what
the perf tests think before landing this though.

Differential Revision: https://phabricator.services.mozilla.com/D76825

BorisChiou committed:
style: Update aspect-ratio syntax for HTML IMG mapped ratio.

Differential Revision: https://phabricator.services.mozilla.com/D76942

emilio committed:
style: Don't use lazy_static for media feature parsing.

This used to be needed when destructors in statics were not allowed, but
we can get rid of it nowadays.

Differential Revision: https://phabricator.services.mozilla.com/D78109

emilio committed:
style: Miscellaneous servo build fixes.

emilio committed:
style: Rustfmt recent changes.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/document.rs
 * @kichjang: components/script/dom/document.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style and script code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 762abba has been approved by `emilio`

<!-- @bors-servo r=emilio 762abbaf9f886fc4463124d88f54fafab0ca1a6d -->

bors-servo commented:
:hourglass: Testing commit 762abbaf9f886fc4463124d88f54fafab0ca1a6d with merge c57e1ec6dda2668942cd7ddd4e33122ab7aeaa1e...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/DvLSX9xBQGyxZkHChO69gg)

emilio committed:
style: Fix some unit tests.

We need to grow dependency by a pointer because of the parent chain for
:is() / :where() unfortunately.

emilio commented:
@bors-servo r+

 * Fixed unit tests.

bors-servo commented:
:pushpin: Commit 16cb510 has been approved by `emilio`

<!-- @bors-servo r=emilio 16cb51097f2681a1c859634a63b2915b18d52cbe -->

bors-servo commented:
:hourglass: Testing commit 16cb51097f2681a1c859634a63b2915b18d52cbe with merge 9b4915e6f5c425afefe7770767d61844e3e1d2f5...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/bR-rXPkWQlKcnX4rljONyg)

emilio committed:
style: fix tidy.

emilio commented:
@bors-servo r+

 * Fixed tidy.

bors-servo commented:
:pushpin: Commit 79c2c44 has been approved by `emilio`

<!-- @bors-servo r=emilio 79c2c447fcdb49758b9c30a2fc0f55a630c5d21f -->

bors-servo commented:
:hourglass: Testing commit 79c2c447fcdb49758b9c30a2fc0f55a630c5d21f with merge 4d95efeebffa0194bd2dde4e5701c4b6d2819f23...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/LN9jJk2qT0uAWoJx4UWvjA)
Approved by: emilio
Pushing 4d95efeebffa0194bd2dde4e5701c4b6d2819f23 to master...
